File: node.py
class Node:
  def __init__(self, *, data, children, parents):
    self.data = data
    self.children = set(children)
    self.parents = set(parents)
    self.visited = False

  def get_all_descendants(self):
    if self is None: return []
    result = [self]
    for child in self.children:
      result.extend(child.get_all_descendants())
    return result

File: utils_ruby/node.rb
class Node
  attr_accessor :data, :children, :visited

  def initialize(data, *children)
    @data = data
    @children = children
    @visited = false
  end

  def get_all_nodes(node=self)
    return node if node.children.empty?
    result = [node]
    for child in node.children do
      result.push(*get_all_nodes(child))
    end
    return result
  end
end
